core/src/main/java/org/infinispan/transaction/impl/TransactionTable.java

```diff
@@ -632,18 +649,52 @@ public CompletedTransactionsInfo() {
        * Once marked as completed (because of commit or rollback) any further prepare received on that transaction are discarded.
        */
-      public void markTransactionCompleted(GlobalTransaction globalTx) {
-         log.tracef("Marking transaction %s as completed", globalTx);
-         completedTransactions.put(globalTx, timeService.time());
+      public void markTransactionCompleted(GlobalTransaction globalTx, boolean successful) {
+         if (trace) log.tracef("Marking transaction %s as completed", globalTx);
+         completedTransactions.put(globalTx, new CompletedTransactionInfo(timeService.time(), successful));
       }
 
       /**
-       * @see #markTransactionCompleted(org.infinispan.transaction.xa.GlobalTransaction)
+       * @see #markTransactionCompleted(GlobalTransaction, boolean)
        */
       public boolean isTransactionCompleted(GlobalTransaction gtx) {
-         if (completedTransactionsInfo == null)
+         if (completedTransactions.containsKey(gtx))
+            return true;
+
+         // Transaction ids are allocated in sequence, so any transaction with a smaller id must have already finished.
+         // Most likely because the prepare command timed out...
+         // Note: We must check the id *after* verifying that the tx doesn't exist in the map.
+         if (gtx.getId() > globalMaxPrunedTxId)
             return false;
+         Long nodeMaxPrunedTxId = nodeMaxPrunedTxIds.get(gtx.getAddress());
+         return nodeMaxPrunedTxId != null && gtx.getId() <= nodeMaxPrunedTxId;
+      }
 
+      public CompletedTransactionStatus getTransactionStatus(GlobalTransaction gtx) {
+         CompletedTransactionInfo completedTx = completedTransactions.get(gtx);
+         if (completedTx != null) {
+            return completedTx.successful ? CompletedTransactionStatus.COMMITTED : CompletedTransactionStatus.ABORTED;
+         }
+
+         // Transaction ids are allocated in sequence, so any transaction with a smaller id must have already finished.
+         // Most likely because the prepare command timed out...
+         // Note: We must check the id *after* verifying that the tx doesn't exist in the map.
+         if (gtx.getId() > globalMaxPrunedTxId)
+            return CompletedTransactionStatus.NOT_COMPLETED;
+         Long nodeMaxPrunedTxId = nodeMaxPrunedTxIds.get(gtx.getAddress());
+         if (nodeMaxPrunedTxId == null) {
+            // We haven't removed any transaction for this node
+            return CompletedTransactionStatus.NOT_COMPLETED;
+         } else if (gtx.getId() > nodeMaxPrunedTxId) {
+            // We haven't removed this particular transaction yet
+            return CompletedTransactionStatus.NOT_COMPLETED;
+         } else {
+            // We already removed the status of this transaction from the completed transactions map
+            return CompletedTransactionStatus.EXPIRED;
+         }
+      }
+
+      public boolean isTransactionStatusLost(GlobalTransaction gtx) {
```

Comment:
not used...
also, if the completed transactions map has the transaction, the status is not lost. or am I missing something?